### PR TITLE
openshift-tuned event-driven change processing

### PR DESCRIPTION
**Pull request by jmencak**
For simplicity, `openshift-tuned` used to process changes in 1s intervals.  While this made the code simpler, it also introduced unnecessary wakeups to process these changes.  These wakeups, of course, cost CPU time and are, therefore, undesirable.

This PR removes the unnecessary wakeups and introduces event-driven processing of these changes.

Other changes:
  - address a potential goroutine leak
  - address a potential case of not-processing Profile object change
  - `s/Tuned/TuneD/` (as per the upstream project) where appropriate
  - logging improvements
  - comments review

### CODE REVIEW

**Comment by jmencak** [comment]
While this is WiP, everything seems to be working fine and the only outstanding issue from my side is whether make the retries of queued events less aggressive.  Thoughts on the current code more than welcome.  Thank you!

/uncc @sjug 
/cc @kpouget

File: pkg/tuned/tuned.go [hunk 1]
```diff
@@ -1044,21 +1103,27 @@ func (c *Controller) changeWatcher() (err error) {
...
 
-		case <-tickerUpdate.C:
-			klog.V(2).Infof("tickerUpdate.C")
+		case <-c.changeCh:
+			var synced bool
+			klog.V(2).Infof("changeCh")
 			if c.daemon.reloading {
-				// Do not reload the Tuned daemon unless it finished reloading
+				// Do not reload the TuneD daemon unless it finished with the previous reload.
+				c.changeChRet <- synced
```
Comment (kpouget) [inline pkg/tuned/tuned.go:1117]: are you relying on Go default bool value for `synced [:= false]` ?
if yes, I find it quite confusing (perhaps that's a common way to proceed in Go?), 
if no then that's a mistake as `synced` isn't explicitly initialized, AFAICU

Comment (kpouget) [reply pkg/tuned/tuned.go:1117]: reading the code outside of the diff, I understand better this line ... but still find it confusing, I would rather write `c.changeChRet <- false` 
and remove the `var synced bool` above

**Review by kpouget — COMMENTED** [review]
thanks for the PR @jmencak, 

I had a look at the code lines and they all makes sense to me,
I didn't try yet to understand overall how it's exactly working, but I can have another round of review tomorrow when the PR moves out of `WIP`

I added a few comments, but mostly only for discussion

I would personally move the comments update into a `trivial` commit,
but I'm not sure if the PR gets squashed before merging or not

**Comment by jmencak** [comment]
> I didn't try yet to understand overall how it's exactly working, but I can have another round of review tomorrow when the PR moves out of `WIP`

Thank you Kevin, much appreciated.  I did find a few other issues when testing that will need to be fixed first, so not sure about "tomorrow". :)

> I added a few comments, but mostly only for discussion
> 
> I would personally move the comments update into a `trivial` commit,
> but I'm not sure if the PR gets squashed before merging or not

I'll definitely squash before merging.

File: pkg/tuned/tuned.go [hunk 2]
```diff
@@ -173,54 +176,56 @@ func newController(stopCh <-chan struct{}) (*Controller, error) {
...
 				}
-				klog.Errorf("unable to sync(%s/%s) reached max retries(%d): %v", workqueueKey.kind, workqueueKey.name, maxRetries, err)
+				klog.Errorf("unable to sync(%s/%s) reached max retries (%d): %v", workqueueKey.kind, workqueueKey.name, maxRetries, err)
 			} else {
 				klog.V(1).Infof("event from workqueue (%s/%s) successfully processed", workqueueKey.kind, workqueueKey.name)
 			}
-			// Successful processing or we're dropping an item after maxRetries unsuccessful retries
-			c.workqueue.Forget(obj)
+			// Successful processing or we're dropping an item after maxRetries unsuccessful retries.
+			c.wqKube.Forget(obj)
```
[the remark this answers is not in the document]
Comment (jmencak) [reply pkg/tuned/tuned.go:228]: I agree this is probably easier to read. Thank you.

File: pkg/tuned/tuned.go [hunk 3]
```diff
@@ -262,10 +268,12 @@ func (c *Controller) sync(key wqKey) error {
 			return err
 		}
 		c.change.profile = true
+		c.wqTuneD.AddRateLimited(wqKey{kind: wqKindDaemon})
```
[the remark this answers is not in the document]
Comment (jmencak) [reply pkg/tuned/tuned.go:271]: The duplication is a bug -- will fix.  Also reconsidering using only `.Add()` and keep `.AddRateLimited()` only for retries.  Need to look into the `workqueues` and their functionality more to decide.  As for the comments, will add them.  Thank you.

File: pkg/tuned/tuned.go (hunk 1 above)
Comment (jmencak) [reply pkg/tuned/tuned.go:1117]: Yes, I rely on the Go default bool value for `synced [:= false]`?  But I also agree that explicit is probably better in this case.  Will change to `c.changeChRet <- false`.

File: pkg/tuned/tuned.go [hunk 4]
```diff
@@ -668,10 +674,62 @@ func (c *Controller) timedUpdater() (err error) {
...
+					c.wqTuneD.AddRateLimited(workqueueKey)
+					return
+				}
+				klog.Errorf("unable to sync(%s/%s) reached max retries (%d)", workqueueKey.kind, workqueueKey.name, maxRetries)
+			} else {
+				klog.V(1).Infof("event from workqueue (%s/%s) successfully processed", workqueueKey.kind, workqueueKey.name)
+			}
+			// Successful processing or we're dropping an item after maxRetries unsuccessful retries.
+			c.wqTuneD.Forget(obj)
+		}()
```
[the remark this answers is not in the document]
Comment (jmencak) [reply pkg/tuned/tuned.go:730]: The point here is to call always call `Done()` once we've finished processing the item.  See: https://github.com/kubernetes/client-go/blob/843bb800b12ac5b3f0d273bd39f889e248b43fed/util/workqueue/queue.go#L145

[the remark this answers is not in the document]
Comment (kpouget) [reply pkg/tuned/tuned.go:730]: got it, so the pattern is something like this
```
func(){
  defer cleanup()
  ...
}()
```
I'll keep it in mind

**Comment by jmencak** [comment]
I believe reviewer comments were addressed at this point. Also fixed a bug when `openshift-tuned` exited unnecessarily when turning on/off debugging.  At this point I do not know of any bugs, however, I believe the workqueue retry aggressiveness still needs to be addressed.

**Comment by jmencak** [comment]
Less aggressive retries implemented, hopefully, this should be all.

**Comment by jmencak** [comment]
Not going to torture with the upgrade testing again until the reviewers agree this is good to go and I squash the PR.

**Comment by kpouget** [comment]
thansk @jmencak, seems good to me!

**Comment by jmencak** [comment]
The only test that this PR could break is "`ci/prow/e2e-aws-operator`"  and that one passed.
/override ci/prow/e2e-upgrade

**Comment by jmencak** [comment]
Bot on PTO?
/joke

**Comment by jmencak** [comment]
@kpouget , thanks for the LGTM comment, `/lgtm` label appreciated whenever you get a time, thanks!

**Pull request merged**